californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -527,302 +620,328 @@ private void handleBlock2Response(final Exchange exchange, final Response respon
+					// if something goes wrong with this blockwise request, e.g. if it times out
+					block.addMessageObservers(request.getMessageObservers());
+					// add an observer that cleans up the block2 transfer tracker if the
+					// block request fails
+					addBlock2CleanUpObserver(block, key);
 
-				// copy options
-				block.setOptions(new OptionSet(request.getOptions()));
-				// make sure NOT to use Observe for block retrieval
-				block.getOptions().removeObserve();
+					status.setCurrentNum(nextNum);
 
-				block.getOptions().setBlock2(szx, m, num);
+					LOGGER.log(Level.FINER, "requesting next Block2 [num={0}]: {1}", new Object[]{ nextNum, block });
+					exchange.setCurrentRequest(block);
+					lower().sendRequest(exchange, block);
 
-				// copy message observers from original request so that they will be notified
-				// if something goes wrong with this blockwise request, e.g. if it times out
-				block.addMessageObservers(request.getMessageObservers());
+				} else {
 
-				responseStatus.setCurrentNum(num);
+					// we have received the last block of the block2 transfer
+
+					LOGGER.log(
+							Level.FINER,
+							"all {0} blocks have been retrieved, assembling response and delivering to application layer",
+							status.getBlockCount());
+					Response assembled = new Response(response.getCode());
+					status.assembleMessage(assembled);
+
+					// set overall transfer RTT
+					assembled.setRTT(System.currentTimeMillis() - exchange.getTimestamp());
+
+					if (status.isNotification()) {
+
+						/*
+						 * When retrieving the rest of a blockwise notification
+						 * with a different token, the additional Matcher state
+						 * must be cleaned up through the call below.
+						 */
+						if (!response.getOptions().hasObserve()) {
+							// call the clean-up mechanism for the additional Matcher entry in exchangesByToken
+							exchange.completeCurrentRequest();
+						}
+					}
 
-				exchange.setCurrentRequest(block);
-				lower().sendRequest(exchange, block);
+					clearBlock2Status(key);
+					LOGGER.log(Level.FINE, "assembled response: {0}", assembled);
+					// Set the original request as current request so that
+					// the Matcher can clean up its state based on the latest
+					// ("current") request's MID and token
+					exchange.setCurrentRequest(exchange.getRequest());
+					// Set the assembled response as current response
+					exchange.setResponse(assembled);
+					upper().receiveResponse(exchange, assembled);
+				}
 
 			} else {
-				LOGGER.log(Level.FINER, "We have received all {0} blocks of the response. Assemble and deliver", responseStatus.getBlockCount());
-				Response assembled = new Response(response.getCode());
 
-				assembleMessage(responseStatus, assembled);
+				// ERROR, wrong block number (server error)
+				// TODO: This scenario is not specified in the draft.
+				// Canceling the request would interfere with Observe, so just ignore it
+				LOGGER.log(Level.WARNING,
```

Comment:
I guess it means that we *do not process* the block returned by the peer. However, sending an RST doesn't make much sense here because the `ReliabiliytLayer` will already have ACKed the CON response. I will remove it ...